### ISSUE TITLE
other actions for Retokenizer?

### ISSUE THREAD

**Issue by bdewilde** [explosion/spaCy#4128]
## Feature description

I'm looking for an efficient way to add data augmentation into my `TextCategorizer` training workflow — namely, by applying a variety of transformations to a given `Doc` and thereby producing a new, similar-but-distinct document. There's a variety of transformations in the literature, but many come down to either swapping two tokens present within a given document/sentence or replacing tokens outright.

Is there any way that the existing `Retokenizer` could incorporate `.replace()` and `.swap()` actions? I appreciate that the resulting sequence of tokens could become ungrammatical, but I don't know what the implications would be for spaCy.

Currently, I'm taking a `Doc` and using its annotations and vectors and such to produce a transformed text string, which I then use to create a new `Doc`, then repeat for the next transformation. It's inefficient. Would be awesome to have some built-in functionality to assist. 🙏 

## Could the feature be a [custom component](https://spacy.io/usage/processing-pipelines#custom-components) or [spaCy plugin](https://spacy.io/universe)?
If so, we will tag it as [`project idea`](https://github.com/explosion/spaCy/labels/project%20idea) so other users can take it on.

Probably! 🙂 tbh I'd like to incorporate the higher-level data augmentation functionality into `textacy`, I'm just hoping for a less hacky / more efficient solution.

**Comment by ines** [comment]
Yay! We're also super interested in data augmentation features and we think that having things like this live in `textacy` (as opposed to the core library) would actually be a great fit.

The main problem I see with adding those retokenization methods is that they're destructive. Currently, no feature, including the retokenizer, breaks the `doc.text == "".join(t.text_with_ws for t in doc)` continuity, which is one of the core principles of spaCy. 

If you need to construct new and modified copies of a `Doc`, the most efficient way to do this would be via the [Cython API](https://spacy.io/api/cython). I know it might sound like a big step to change `textacy` to use Cython, but it would open up a whole lot of new possibilities, and we'd be happy to help you with getting everything set up.

**Comment by bdewilde** [comment]
Glad to hear you're interested! I wondered if `doc.text != "".join(t.text_with_ws for t in doc)` would be a sticking point, so I totally understand preferring such functionality elsewhere. I'm happy to give it a go in `textacy`. 👍 

I've hacked something together that's more performant than before but still feels clumsy — currently blocked on the problem of getting good synonyms for words, though I suppose that's a separate issue 😅 — but you're saying that changing the underlying tokenization is doable via spaCy's Cython API? I've resisted cythonizing parts of textacy because of the potential for user installation issues and added complexity to the development process, but if you could point me to any guides/guidelines that have worked for y'all, then I'd be willing to try!

**Comment by ines** [comment]
> I've hacked something together that's more performant than before but still feels clumsy

How are you currently doing it? Are you using `Doc.to_array` / `Doc.from_array` and then swapping the values?

> but you're saying that changing the underlying tokenization is doable via spaCy's Cython API?

Yes, definitely! @honnibal can probably write a quick code snippet for you that shows the idea.

> I've resisted cythonizing parts of textacy because of the potential for user installation issues and added complexity to the development process

Yeah, I know what you mean. It's definitely doable, though. You need one additional step in your `setup.py` that compiles the Cython modules, which you can probably copy from this repo.

If you want the smooth no-compiler-required installation with `pip`, you will need to build and upload wheels in addition to the sdist. This is slightly more work and the "tricky part". But since we need all of this as well, we've already solved that problem and open-sourced the solution as [`wheelwright`](https://github.com/explosion/wheelwright). So you could fork that, add your Travis / Appveyor tokens and then have it build wheels for all Python version / OS combinations (except for 2.7 on Windows, I think – but spaCy doesn't have that either).

**Comment by honnibal** [comment]
It might be helpful to have a look at the `Doc` definition here: https://github.com/explosion/spaCy/blob/master/spacy/tokens/doc.pxd

The key part is really just:

```cython
cdef class Doc:
    cdef Pool mem
    cdef TokenC* c
    cdef int length
```

So, the `Doc` object is basically a C array, of some number of `TokenC` structs. Here's some example code that replaces one word with another:

```cython
cdef int replace_token(Doc doc, int i, attr_t orth) except -1:
    """Replace a word with another, fetching the underlying LexemeC* pointer from the Vocab."""
    old_length = doc.c[i].lex.length
    doc.c[i].lex = doc.vocab.get_by_orth(doc.vocab.mem, orth)
    # Adjust the character start indices of subsequent tokens.
    offset = doc.c[i].lex.length - old_length
    for j in range(i+1, doc.length):
        doc.c[j].idx += offset
```

Naively, swapping two tokens `i` and `j` would be: `doc.c[i].lex, doc.c[j].lex = doc.c[j].lex, doc.c[i].lex`. However, you'd need to adjust the start indices to line everything up.

Bad things will happen if you end up getting the parse tree into an inconsistent state, so that's also something to keep in mind if you're messing with the C representation. For instance, if you delete a token and the parse is set, the head offsets will be wrong, and the left/right children information will be wrong too. Otherwise, it should be pretty easy.

If you want to create a copy of a `Doc`, something like this should be the quickest way:

```cython
from libc.string cimport memcpy
from spacy.tokens.doc cimport Doc
import copy

def copy_doc(Doc doc):
    cdef Doc new_ = Doc(doc.vocab)
    new_._realloc(doc.max_length)
    memcpy(new_.c, doc.c, doc.max_length * sizeof(doc.c[0]))
    new_.cats = dict(doc.cats)
    new_.is_tagged = doc.is_tagged
    new_.is_parsed = doc.is_parsed
    new_.sentiment = doc.sentiment
    new_.user_data = copy.deepcopy(doc.user_data)
    new_.user_token_hooks = dict(doc.user_token_hooks)
    new_.user_span_hooks = dict(doc.user_span_hooks)
    new_.user_hooks = dict(doc.user_hooks)
```

We should probably have that function in the library, come to think of it...

**Comment by bdewilde** [comment]
Alright, here's my first pass on implementing basic text data augmentation in textacy: https://github.com/chartbeat-labs/textacy/pull/268. It's... not great, to be honest, but I wanted to push something out sooner than later, then improve it as needed. I'm sticking a pin in cython-izing textacy for just a bit longer — I'll get there! just nervous about diving in over my head 😅 — but thank you for pointing me in promising directions.

**Comment by honnibal** [comment]
@bdewilde congrats on the release, and glad you got the data augmentation implemented! It would be cool in future to try out a transformer model for smarter replacements. I could imagine that working well.
